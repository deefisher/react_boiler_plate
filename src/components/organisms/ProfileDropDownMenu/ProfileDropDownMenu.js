import { Box, DropButton, Text } from 'grommet';
import { FormDown, Logout, SettingsOption, User } from 'grommet-icons';
import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { routePaths } from '../../../utils/routing/routePaths';

export const labelText = {
    SETTINGS: 'Settings',
    LOGOUT: 'Log out',
};

export const ariaLabels = {
    dropdownMenu: 'Dropdown menu',
    dropMenu: 'Drop menu',
    profileDropdown: 'Profile drop down',
};

const profileDropMenuItems = [
    {
        label: labelText.SETTINGS,
        icon: <SettingsOption size="15px" />,
        href: routePaths.settings,
        onItemClick: () => {},
    },
    { label: labelText.LOGOUT, icon: <Logout size="15px" />, onItemClick: () => {} },
];

const StyledNoDecorationLink = styled(Link)`
    text-decoration: none;
    color: black;
`;

const StyledNoBoxShodowBox = styled(Box)`
    box-shadow: none;
`;

const DropMenu = ({ items, ...props }) => {
    return (
        <Box background="white" a11yTitle={ariaLabels.dropdownMenu}>
            {items.map((item, i) => (
                <StyledNoDecorationLink to={item.href ? item.href : '#'} key={`drop-item-${i}`}>
                    <StyledNoBoxShodowBox
                        a11yTitle={ariaLabels.dropMenu}
                        hoverIndicator={{ background: 'light-3' }}
                        onClick={item.onItemClick}
                    >
                        <Box pad={{ vertical: '1rem', horizontal: 'medium' }} width={{ min: '9rem' }}>
                            <Box direction={'row'} align="center">
                                {item.icon}
                                <Box margin={{ left: 'small' }}>
                                    <Text size="small">{item.label}</Text>
                                </Box>
                            </Box>
                        </Box>
                    </StyledNoBoxShodowBox>
                </StyledNoDecorationLink>
            ))}
        </Box>
    );
};

const ProfileDropDownMenu = (props) => {
    return (
        <DropButton
            {...props}
            a11yTitle={ariaLabels.profileDropdown}
            label={
                <Box direction="row" align="center">
                    <Box
                        justify="center"
                        align="center"
                        round="50%"
                        background="brand"
                        width="35px"
                        height="35px"
                        margin={{
                            right: 'small',
                        }}
                    >
                        <User size="19px" color="white" />
                    </Box>
                    <FormDown />
                </Box>
            }
            dropAlign={{
                top: 'bottom',
                right: 'right',
            }}
            dropContent={<DropMenu items={profileDropMenuItems} />}
            dropProps={{
                elevation: 'custom',
            }}
            plain
        />
    );
};

export default ProfileDropDownMenu;

DropMenu.propTypes = {
    items: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string,
            icon: PropTypes.element,
            onItemClick: PropTypes.func,
        }),
    ).isRequired,
};
