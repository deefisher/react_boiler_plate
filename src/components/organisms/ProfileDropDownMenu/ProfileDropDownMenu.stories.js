import { Box } from 'grommet';
import React from 'react';
import ProfileDropDownMenu from './ProfileDropDownMenu';

export const DefaultProfileDropDownMenu = (props) => <ProfileDropDownMenu {...props} />;
export const OpenProfileDropDownMenu = () => <DefaultProfileDropDownMenu open />;
export const ProfileDropDownMenuInDark = () => (
    <Box background="topbar" width="300px" pad="medium">
        <DefaultProfileDropDownMenu open />
    </Box>
);

const ProfileDropDownMenuStory = {
    title: 'Organisms/ProfileDropDownMenu',
    component: ProfileDropDownMenu,
};

OpenProfileDropDownMenu.args = {
    with: 'props',
};
OpenProfileDropDownMenu.parameters = {
    // Sets the delay for a specific story.
    chromatic: { delay: 1000 },
};

export default ProfileDropDownMenuStory;
