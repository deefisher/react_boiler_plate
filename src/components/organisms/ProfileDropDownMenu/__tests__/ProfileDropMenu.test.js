import { act, fireEvent, render, screen } from '@testing-library/react';
import ProfileDropDownMenu, { ariaLabels, labelText } from '../ProfileDropDownMenu';
import { MemoryRouter } from 'react-router-dom';

describe('ProfileDropMenu', () => {
    let wrapper;
    window.scrollTo = jest.fn();
    beforeEach(() => {
        wrapper = render(<ProfileDropDownMenu />, { wrapper: MemoryRouter });
    });

    it('should render ProfileDropMenu correctly', () => {
        const { asFragment } = wrapper;
        expect(asFragment()).toMatchSnapshot();
    });

    it('should open drop menu on click', () => {
        expect(screen.queryByText(labelText.SETTINGS)).not.toBeTruthy();

        act(() => {
            fireEvent.click(screen.getByLabelText(ariaLabels.profileDropdown));
        });

        expect(screen.queryByText(labelText.SETTINGS)).toBeTruthy();
    });
});

//TODO add click on settings as taiko test
