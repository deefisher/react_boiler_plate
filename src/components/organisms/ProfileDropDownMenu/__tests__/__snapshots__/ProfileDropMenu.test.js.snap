// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileDropMenu should render ProfileDropMenu correctly 1`] = `
<DocumentFragment>
  <button
    aria-label="Profile drop down"
    class="StyledButton-sc-323bzc-0 kVhqUm"
    type="button"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hznAXC"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 gKVznV"
      >
        <svg
          aria-label="User"
          class="StyledIcon-sc-ofa7kd-0 gYDKuB"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 24v-5m8 5v-5M3 24v-5c0-4.97 4.03-8 9-8s9 3.03 9 8v5m-9-13a5 5 0 1 0 0-10 5 5 0 0 0 0 10z"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
      <svg
        aria-label="FormDown"
        class="StyledIcon-sc-ofa7kd-0 jjvRzN"
        viewBox="0 0 24 24"
      >
        <path
          d="m18 9-6 6-6-6"
          fill="none"
          stroke="#000"
          stroke-width="2"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;
